src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs

```diff
@@ -0,0 +1,375 @@
+
+        public const string StackTypeName = "System.Collections.Stack";
+        public const string StackGenericTypeName = "System.Collections.Generic.Stack`1";
+
+        public const string QueueTypeName = "System.Collections.Queue";
+        public const string QueueGenericTypeName = "System.Collections.Generic.Queue`1";
+
+        public const string SetGenericInterfaceTypeName = "System.Collections.Generic.ISet`1";
+        public const string SortedSetGenericTypeName = "System.Collections.Generic.SortedSet`1";
+        public const string HashSetGenericTypeName = "System.Collections.Generic.HashSet`1";
+
+        public const string LinkedListGenericTypeName = "System.Collections.Generic.LinkedList`1";
+
+        public const string DictionaryInterfaceTypeName = "System.Collections.IDictionary";
+        public const string DictionaryGenericTypeName = "System.Collections.Generic.Dictionary`2";
+        public const string DictionaryGenericInterfaceTypeName = "System.Collections.Generic.IDictionary`2";
+        public const string ReadOnlyDictionaryGenericInterfaceTypeName = "System.Collections.Generic.IReadOnlyDictionary`2";
+        public const string SortedDictionaryGenericTypeName = "System.Collections.Generic.SortedDictionary`2";
+        public const string KeyValuePairGenericTypeName = "System.Collections.Generic.KeyValuePair`2";
+
+        public const string ArrayListTypeName = "System.Collections.ArrayList";
+
+        // In the order we wish to detect a derived type.
+        private static readonly Type[] s_genericInterfacesWithAddMethods = new Type[]
+        {
+            typeof(IDictionary<,>),
+            typeof(ICollection<>),
+        };
+
+        // In the order we wish to detect a derived type.
+        private static readonly Type[] s_nonGenericInterfacesWithAddMethods = new Type[]
+        {
+            typeof(IDictionary),
+            typeof(IList),
+        };
+
+        // In the order we wish to detect a derived type.
+        private static readonly Type[] s_genericInterfacesWithoutAddMethods = new Type[]
+        {
+            typeof(IReadOnlyDictionary<,>),
+            typeof(IReadOnlyCollection<>),
+            typeof(IReadOnlyList<>),
+            typeof(IEnumerable<>),
+        };
+
+        public static Type GetImplementedCollectionType(Type queryType)
+        {
+            Debug.Assert(queryType != null);
+
+            if (queryType.IsAbstract ||
+                queryType.IsInterface ||
+                !(typeof(IEnumerable).IsAssignableFrom(queryType)) ||
+                queryType == typeof(string) ||
+                queryType.IsArray ||
+                queryType == typeof(object) ||
+                IsNativelySupportedCollection(queryType))
+            {
+                return queryType;
+            }
+
+            Type baseType = queryType.GetTypeInfo().BaseType;
+
+            // Check if the base type is a supported concrete collection.
```

Comment:
Is it worthwhile to do this recursively? (any base class)?

Reply:
I don't think so. We'd still need to check for implemented interfaces, so not having any base class won't tell us much.

If we want to support detection of an implemented concrete class at a higher level than a types parent class, then we could recurse to the base type (if any) at the last line in this method. I'd hold of on this until there's a demonstrated user need though.